### ISSUE TITLE
Name deprecations do not unbold names in cached matrix boxes on observation indexes

### ISSUE THREAD

**Issue by JoeCohen** [MushroomObserver/mushroom-observer#1907]
Sometimes light (correct), but sometimes bold (wrong).
Examples: 
- https://mushroomobserver.org/observations?pattern=name%3A%22Trichia+decipiens%22
- goto https://mushroomobserver.org/names/21206 
In the "About this taxon" panel, click "this name"

**Comment by nimmolo** [comment]
@JoeCohen @pellaea 
To be honest, I don't understand what governs the boldness of name display. Is it `Textile.register_name`?

**Comment by nimmolo** [comment]
Ah, I don't believe it's a Textile issue. It seems "Trichia decipiens" may be a very recently deprecated name, because in my db snapshot, it is accepted. If that's the case, then previously cached matrix_boxes will have it bolded, and matrix boxes pulled after the deprecation will have it not bolded. 

If it's worth fixing, we need to alter the controller action for name deprecations so that it "touches" all associated observations. In other words, the `deprecate` action needs to query all obs where the name is the consensus id, and `touch`, which changes the `updated_at` column on the obs.

For an observation matrix box, the cache only keeps track of the AR records of **the observation, its namings and votes**. Changes to other records associated with the observation will not currently invalidate the cache of the matrix box. 

In other words, anything else in the matrix box that we want to keep updated (like some kind of location name change, or user "login" change) will need to be altered so that the update actions explicitly `touch` the observation.

This may sound like a pain in the butt, but nothing has sped up our home page more than this caching. Maintaining that improvement will require changes like this.

**Comment by nimmolo** [comment]
Thanks for checking it out, @mo-nathan. But heads up: invalidating the entire cache is much costlier than touching 1000 records in terms of server load. That's what I'm learning from all these experiments to speed up the site. The "way to just invalidate the relevant cache entries" is to `touch` them, it's not an expensive operation.

I want to be sure everyone has an appreciation of HOW MANY lookups and renders the cache is saving us. If you haven't examined your dev log, i will do a little TED-talk about it at the next meeting with show and tell.

**Comment by nimmolo** [comment]
@JoeCohen - good question. 
Yes is the answer. All names whose status has changed will need to go into one batch in the query, and then they get `touched` in a single batch operation. 

That's why I nominate you to handle the PR 😁, you're more aware of Name logic. I also think everyone needs some practice writing cache- and query-friendly code.

Check the `Names::SynonymsController`: 
Out of a given list of synonyms, the admin picks a winner, and it changes place with some other name. But other names may remain deprecated, so there's no change in their status. Then, the controller goes through the list, and names are approved/deprecated one by one in separate transactions.

But in order to make our `touch` operation quick, we need to send the names in a single batch (of AR records, NOT IDs). So we need to assemble an array of the names that have actually changed. 

Then:
```ruby
# EVERYONE: please note an important recent discovery here:
# The way to produce a much, much quicker AR subquery is to 
# pass an array of ** records ** (not IDs!) to an ID field
Observation.where(name_id: changed_names).touch
```

**Comment by mo-nathan** [comment]
I looked at this a bit.  The first thing I discovered is that my local is not storing anything in the cache so I cannot reproduce the bug.  The database exists and the migrations have run, but there is nothing in there.  Is there some other config that needs to be done to get it to happen when in development mode?

I am not at all convinced that touching every observation associated with a name is the right or practical thing to do.  In many cases that implies searching for and changing 100s of observations and technically the object itself did not get updated so in my view changing the updated_at field is not appropriate.  What is the problem with associating a cached matrix_box with it's name's updated_at field or some such?  That just seems sensible.

I spent some time trying to figure out how this cache works and I can't figure out what gets cached where and why.  In my opinion, this is something @nimmolo introduced, so it's his responsibility to fix it.  The reality is doing caching correctly is HARD.  Yes, if done correctly it can have huge upsides, but if it is done wrong it breaks stuff as this has twice now.  If I'm required to fix it, at this point I will turn it off since that's the fastest way I see to restore the expected functionality so I can get back to the work that I feel is most important for the success of MO.  At the moment that is figuring out how to build the mobile app which is forcing me to learn multiple new technologies.  Adding an unrelated caching system on top of that is not the best use of my time.  I appreciate the work @nimmolo has done on performance and it's great that he thinks that is the most important thing he can do for the success of MO, but it is not fair to assume that all of us are going to drop what we're doing to work on a bug that he has introduced.

I did do some looking into the resources the cache is using in production.  It has been running since Feb. 5 and there are 378,947 cache entries.  Nothing has come out of the cache yet and it is taking up 979.6MB of disk space.  From what I can see in the documentation it starts dropping things after 2 weeks using a First In, First Out algorithm, so it should end up maxing out around 1.5GB.  We currently have 12G of disk on the production server, so we could easily double or even quadruple the cache with no problems.  A factor of 10 would be a problem, but it looks like we can also cap it by total size if needed.

**Comment by nimmolo** [comment]
@mo-nathan — Thanks for taking the time to investigate the caching. I hear you that it's not working for you locally; I'll look for where you might be missing a config, I think I have an idea - I'll put it in Slack if I find it.

Regarding this:
> I am not at all convinced that touching every observation associated with a name is the right or practical thing to do. In many cases that implies searching for and changing 100s of observations and technically the object itself did not get updated so in my view changing the updated_at field is not appropriate. What is the problem with associating a cached matrix_box with it's name's updated_at field or some such? That just seems sensible.

When we were experimenting with getting it working, Jason and I discussed how best to invalidate the `Observation` caches when a `Naming` is changed. We had a choice of how to do it, because touching any column on the Obs will invalidate caches of that Obs. That discussion is relevant here (it would have been appropriate for a full meeting, but anyway...): 
- In most cases on MO, we are sorting observations by `log_updated_at`, a separate column which has a higher threshold for change. We added that column back when we switched the home page from RssLogs to Observations. It was needed because `updated_at` happens much more often than `log_updated_at`, and trivial updates were inappropriately bumping obs to the top of the feed.
- It seemed to me that a `naming` change didn't qualify as a common-sense `update` change to the Observation. I had brought up the idea of adding a new column like `cache_updated_at`
- Jason pointed out that `updated_at` is a column we do not really use for Observations anyway, so it seems like an OK column to touch (as well as being automatically connected to the `touch` method via AR). He also pointed out that a change to Naming is in a sense a material change to the Observation.

I feel like the same logic could apply to a Name deprecation. 

You also mentioned the cost of the `touch_all` query, and I believe your impression of the relative cost is off. I only know because i've been working with this stuff a lot. Try running this locally in your console: 
```ruby
blue = Name.find_by(text_name: "Lepista nuda")
Observation.where(name_id: blue.synonyms).touch_all
```
_Lepista nuda_ has 6 synonyms, so this is not a trivial example. Running this touches 1170 observations in my db snapshot, and it's done in under a second. 

By contrast, flushing the entire cache of matrix_boxes would temporarily add ~2 seconds to the db load of nearly every commonly-hit observation index or search result, even potentially cached several pages deep, plus the indexes of obs needing id's. And that query time does not include the increased render time to execute the erb templates. It's more costly in several ways.

Deprecating names is not nearly as common a task as viewing an index, so it seems fair to me to put the work in that action.

**What we cache:**
As I mentioned to Joe in another thread, to find everything we cache, it's easy. You just have to search for the command `cache` in the helpers and views. It's only in there a few places. We cache `matrix_box` partials of `observations` and `rss_logs` (only), and we cache the `left_nav` in several separate fragments.

I don't think it's really that hard to pull off this limited caching, when everybody's aware of how it works. Of all the noticeable changes i've done, this was by far the easiest, and it would still count as a breeze if we discover dozens of necessary tweaks like this one, which I don't believe we will. 

Understanding Rails caching will require due diligence from everybody, though, just as learning MO's idiosyncratic code has taken a ton of study on my part. Luckily this type of caching has been around for 10 years, there's a wealth of documentation and Q&A's, and it's much easier to get the hang of. 

Once your local Rails caching is working, I believe you'll start to get a feel for the cost of all this stuff as it's logged locally. We're barely caching any HTML yet, and we're getting huge benefits in render time. But getting spendy with the cache at this point is unwise in my opinion. It's silly to fill up our finite space with duplicate copies of nearly identical stuff when we may want to cache Observation "show" partials and maybe Name views once the matrix boxes are ironed out. Treating queries and rendering as "free" is what got us into the current slow state. Caching is all about learning the cost of "free".

**Comment by mo-nathan** [comment]
That's lovely.  However, you are missing the point of what I said.  Please read the third paragraph of what I wrote again.  Either fix the bug or I will turn off the cache this evening.  I will be happy to turn it back on once the bug is fixed.  The most precious commodity we have are volunteer developer hours.  I really appreciate all the hours you have put in, but I don't ever expect that anyone will put in more time at my whim.  You cannot expect that other people will be able to instantly make time to fix bugs that you introduce in the way that you think it should be done.  If I had all the time in the world I would love to learn all the stuff you describe.  However, I have major problems going on at work that take priority over MO.  I look forward to reviewing your PR.

**Comment by nimmolo** [comment]
I've done some experiments with [the "Russian Doll" pattern of nested cached fragments](https://guides.rubyonrails.org/caching_with_rails.html#russian-doll-caching), and I found out a few things:

1. For russian doll caching to work, the inner nested record (here, the Name) must be made to `touch` the outer record (the Obs) when it's updated, or it doesnt invalidate the cache either. For us, updating associated Obs for every possible update of a Name seems a bit much. Synonymy changes still seem relevant, to me. 
2. Russian Doll is more useful when the nested HTML is a reusable component, like a whole Name "panel" that might appear outside the matrix box context. For our "inner dolls", we'd only be caching the `<a>` links to the name/location/user. Caching these as separate nested (reusable) fragments seems like overkill, considering that even if we do that, we still have to `touch` the obs when a change is relevant. It would be easier to keep the matrix box as a single cached fragment. 
3. I believe I found (with Jason's help) the right spot in the code to touch associated Obs for Name synonymy/author changes, and for User profile changes (such as a user changing their login handle). But Location methods for merges in particular seem confused. I need some help figuring those out.
